lib/ivis_opengl/pielighting.cpp

```diff
@@ -0,0 +1,199 @@
+/*
+	This file is part of Warzone 2100.
+	Copyright (C) 1999-2004  Eidos Interactive
+	Copyright (C) 2005-2020  Warzone 2100 Project
+
+	Warzone 2100 is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+
+	Warzone 2100 is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with Warzone 2100; if not, write to the Free Software
+	Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+*/
+
+#include "pielighting.h"
+#include "gfx_api.h"
+#include <array>
+#include <glm/glm.hpp>
+#include <algorithm>
+#include <functional>
+#include "culling.h"
+#include "src/profiling.h"
+#include "lib/framework/debug.h"
+
+
+PIELIGHT& LightMap::operator()(int32_t x, int32_t y)
+{
+	// Clamp x and y values to actual ones
+	// Give one tile worth of leeway before asserting, for units/transporters coming in from off-map.
+	ASSERT(x >= -1, "mapTile: x value is too small (%d,%d) in %dx%d", x, y, mapWidth, mapHeight);
+	ASSERT(y >= -1, "mapTile: y value is too small (%d,%d) in %dx%d", x, y, mapWidth, mapHeight);
+	x = MAX(x, 0);
+	y = MAX(y, 0);
+	ASSERT(x < mapWidth + 1, "mapTile: x value is too big (%d,%d) in %dx%d", x, y, mapWidth, mapHeight);
+	ASSERT(y < mapHeight + 1, "mapTile: y value is too big (%d,%d) in %dx%d", x, y, mapWidth, mapHeight);
+	x = MIN(x, mapWidth - 1);
+	y = MIN(y, mapHeight - 1);
+
+	return data[x + (y * mapWidth)];
+}
+
+PIELIGHT& LightMap::operator()(Vector2i const& v)
+{
+	return operator()(v.x, v.y);
+}
+
+void LightMap::reset(size_t width, size_t height)
+{
+	mapWidth = static_cast<int32_t>(width);
+	mapHeight = static_cast<int32_t>(height);
+	data = std::make_unique<PIELIGHT[]>(width * height);
+	ASSERT(data != nullptr, "Out of memory");
+}
+
+
+static LightingData scene;
+
+LightingData& getCurrentLightingData()
+{
+	return scene;
+}
+
+namespace {
+	auto getLightBoundingBox(const LIGHT& light)
```

Comment:
Try to avoid having function signatures annotated with `auto` alone. IMO, the only two justified use-cases are:
* Avoid too much typing for return type
* Generic code with template-dependent names for return types

In other cases it provides no convenience and moreover, requires one to look into the implementation to actually find out the return type.

In this particular case you can substitute `BoundingBox` without sacrificing clarity.